pallets/randomness/src/vrf.rs

```diff
@@ -0,0 +1,122 @@
+/// Called in previous block's `on_finalize`
+pub(crate) fn set_input<T: Config>() {
+	let input = T::VrfInputGetter::get_vrf_input();
+	if let Some(last_vrf_input) = <CurrentVrfInput<T>>::take() {
+		// logs if input uniqueness assumptions are violated (no reuse of vrf inputs)
+		if last_vrf_input.storage_root == input.storage_root
+			|| last_vrf_input.slot_number == input.slot_number
+		{
+			log::warn!(
+				"VRF on_initialize: storage root or slot number did not change between \
+			current and last block. Nimbus would've panicked if slot number did not change \
+			so probably storage root did not change."
+			);
+		}
+	}
+	<CurrentVrfInput<T>>::put(input);
+}
+
+/// Returns weight consumed in `on_initialize`
+pub(crate) fn set_output<T: Config>() -> Weight {
+	// Do not set the output in the first block (genesis or runtime upgrade)
+	// because we do not have any input for author to sign (which would be set in last block)
+	if <NotFirstBlock<T>>::get().is_none() {
+		<NotFirstBlock<T>>::put(());
+		return T::DbWeight::get().read + T::DbWeight::get().write;
+	}
+	let input = <CurrentVrfInput<T>>::get().expect("VrfInput must be set to verify VrfOutput");
+	let mut block_author_vrf_id: Option<VrfId> = None;
+	let PreDigest {
+		vrf_output,
+		vrf_proof,
+	} = <frame_system::Pallet<T>>::digest()
+		.logs
+		.iter()
+		.filter_map(|s| s.as_pre_runtime())
+		.filter_map(|(id, mut data)| {
+			if id == VRF_ENGINE_ID {
+				if let Ok(vrf_digest) = PreDigest::decode(&mut data) {
+					Some(vrf_digest)
+				} else {
+					panic!("failed to decode VRF PreDigest");
+				}
+			} else {
+				if id == NIMBUS_ENGINE_ID {
+					let nimbus_id = NimbusId::decode(&mut data)
+						.expect("NimbusId encoded in pre-runtime digest must be valid");
+
+					block_author_vrf_id = Some(
+						T::VrfKeyLookup::lookup_keys(&nimbus_id)
+							.expect("No VRF Key Mapped to this NimbusId"),
+					);
+				}
+				None
+			}
+		})
+		.next()
+		.expect("VRF PreDigest was not included in the digests (check rand key is in keystore)");
+	let block_author_vrf_id =
+		block_author_vrf_id.expect("VrfId encoded in pre-runtime digest must be valid");
+	let pubkey = schnorrkel::PublicKey::from_bytes(block_author_vrf_id.as_slice())
+		.expect("Expect VrfId to be valid schnorrkel public key");
+	let transcript = make_transcript::<T::Hash>(input.slot_number, input.storage_root);
+	// Verify VRF output + proof using input transcript and VrfId
+	assert!(
+		pubkey
+			.vrf_verify(transcript.clone(), &vrf_output, &vrf_proof)
+			.is_ok(),
+		"VRF signature verification failed"
+	);
+	let vrf_output: Randomness = vrf_output
+		.attach_input_hash(&pubkey, transcript)
+		.ok()
+		.map(|inout| inout.make_bytes(&VRF_INOUT_CONTEXT))
+		.expect("VRF output encoded in pre-runtime digest must be valid");
+	let raw_randomness_output = T::Hash::decode(&mut &vrf_output[..]).ok();
+	LocalVrfOutput::<T>::put(raw_randomness_output);
+	// Supply randomness result
```

Comment:
```suggestion
    // If we have any "Local" RandomnessResults for this block (meaning we have at least one such
    // request) then provide our new randomness for them
```

Reply:
I basically rephrased your answer to one of my questions above and replaced this comment with it